### PR TITLE
[Merged by Bors] - feat(fun_prop): failure cache for `fun_prop` and command to print out `fun_prop` theorems

### PR DESCRIPTION
**Pull request by lecopivo**
Add new cache to `fun_prop`  for failed goals. `fun_prop` used to take exponentially long to fail without this cache.
Also add command `#print_fun_prop_theorems HAdd.hAdd`  prints out all `fun_prop` theorems associated with addition.

---

Command useful for maintenance of `fun_prop` attribute. Allows you to to see all the theorems for a given function thus spotting duplicated and bad theorems more easily.

This will also be useful on deciding priority for new `fun_prop` theorems once priority is supported.

### CODE REVIEW

File: MathlibTest/fun_prop_dev.lean [hunk 1]
```diff
@@ -553,3 +554,49 @@ example [Add α] (y : α):
...
+example (f : R → R) (hf : Con f) :
+    Con (fun x ↦ (f (x + 3)) + 2) := by fun_prop -- succeeds in 6ms
+example (f : R → R) (hf : Con f) :
+    Con (fun x ↦ (f (x + 3)) + (2 + f (x + 1))) := by fun_prop -- succeeds in 8ms
+example (f : R → R) (hf : Con f) :
+    Con (fun x ↦ (f (x + 3)) + (2 + f (x + 1)) + x) := by fun_prop -- succeeds in 10ms
+example (f : R → R) (hf : Con f) :
+    Con (fun x ↦ (f (x + 3)) + 2 + f (x + 1) + x + 1) := by fun_prop -- succeeds in 11ms
+
+-- this used to fail in exponentially increasing time, up to 6s for the last example
```
Comment (fpvandoorn) [inline MathlibTest/fun_prop_dev.lean:580]: Could you turn these into actual tests? (e.g. by using `fail_if_success`/`success_if_fail_with_msg` + some low timeout (whose failure is not caught by the `fail_if_success`)

File: Mathlib/Tactic/FunProp/Types.lean [hunk 2]
```diff
@@ -97,8 +97,10 @@ structure Context where
 /-- `fun_prop` state -/
 structure State where
   /-- Simp's cache is used as the `fun_prop` tactic is designed to be used inside of simp and
-  utilize its cache -/
+  utilize its cache. It holds succesfull goals. -/
```
Comment (fpvandoorn) [inline Mathlib/Tactic/FunProp/Types.lean:100]:
```suggestion
  utilize its cache. It holds successful goals. -/
```

**Review by fpvandoorn — APPROVED** [review]
Great! LGTM

bors d+

**Comment by fpvandoorn** [comment]
Oh, please update PR description to include the failure cache.

**Comment by fpvandoorn** [comment]
Did you see I also left 2 review comments?

**Comment by lecopivo** [comment]
Snap I missed that, I just installed the GitHub mobile app and I'm all confused by it.

**Comment by fpvandoorn** [comment]
Feel free to cancel, or do a tiny follow-up PR.

**Comment by lecopivo** [comment]
How do I cancel?

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**